### ISSUE TITLE
Security policy always uses 'path.data' even when node.data:false

### ISSUE THREAD

**Issue by synhershko** [elastic/elasticsearch#27572]
I have a client node installed via apt-get install and configured as such:

```
node.master: false
node.data: false
node.ingest: false
```

When I try to run it (`sudo service elasticsearch start`) I'm getting the following exception in the logs:

```
[2017-11-28T19:36:06,220][ERROR][o.e.b.Bootstrap          ] Exception
java.lang.IllegalStateException: Unable to access 'path.data' (/usr/share/elasticsearch/data)
        at org.elasticsearch.bootstrap.FilePermissionUtils.addDirectoryPath(FilePermissionUtils.java:70) ~[elasticsearch-6.0.0.jar:6.0.0]
        at org.elasticsearch.bootstrap.Security.addFilePermissions(Security.java:287) ~[elasticsearch-6.0.0.jar:6.0.0]
        at org.elasticsearch.bootstrap.Security.createPermissions(Security.java:242) ~[elasticsearch-6.0.0.jar:6.0.0]
        at org.elasticsearch.bootstrap.Security.configure(Security.java:120) ~[elasticsearch-6.0.0.jar:6.0.0]
        at org.elasticsearch.bootstrap.Bootstrap.setup(Bootstrap.java:207) ~[elasticsearch-6.0.0.jar:6.0.0]
        at org.elasticsearch.bootstrap.Bootstrap.init(Bootstrap.java:322) [elasticsearch-6.0.0.jar:6.0.0]
        at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:130) [elasticsearch-6.0.0.jar:6.0.0]
        at org.elasticsearch.bootstrap.Elasticsearch.execute(Elasticsearch.java:121) [elasticsearch-6.0.0.jar:6.0.0]
        at org.elasticsearch.cli.EnvironmentAwareCommand.execute(EnvironmentAwareCommand.java:69) [elasticsearch-6.0.0.jar:6.0.0]
        at org.elasticsearch.cli.Command.mainWithoutErrorHandling(Command.java:134) [elasticsearch-6.0.0.jar:6.0.0]
        at org.elasticsearch.cli.Command.main(Command.java:90) [elasticsearch-6.0.0.jar:6.0.0]
        at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:92) [elasticsearch-6.0.0.jar:6.0.0]
        at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:85) [elasticsearch-6.0.0.jar:6.0.0]
Caused by: java.nio.file.AccessDeniedException: /usr/share/elasticsearch/data
        at sun.nio.fs.UnixException.translateToIOException(UnixException.java:84) ~[?:1.8.0_151]
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102) ~[?:1.8.0_151]
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107) ~[?:1.8.0_151]
        at sun.nio.fs.UnixFileSystemProvider.createDirectory(UnixFileSystemProvider.java:384) ~[?:1.8.0_151]
        at java.nio.file.Files.createDirectory(Files.java:674) ~[?:1.8.0_151]
        at java.nio.file.Files.createAndCheckIsDirectory(Files.java:781) ~[?:1.8.0_151]
        at java.nio.file.Files.createDirectories(Files.java:767) ~[?:1.8.0_151]
        at org.elasticsearch.bootstrap.Security.ensureDirectoryExists(Security.java:401) ~[elasticsearch-6.0.0.jar:6.0.0]
        at org.elasticsearch.bootstrap.FilePermissionUtils.addDirectoryPath(FilePermissionUtils.java:68) ~[elasticsearch-6.0.0.jar:6.0.0]
        ... 12 more
[2017-11-28T19:36:06,247][WARN ][o.e.b.ElasticsearchUncaughtExceptionHandler] [my-cluster-client000000] uncaught exception in thread [main]
org.elasticsearch.bootstrap.StartupException: java.lang.IllegalStateException: Unable to access 'path.data' (/usr/share/elasticsearch/data)
        at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:134) ~[elasticsearch-6.0.0.jar:6.0.0]
        at org.elasticsearch.bootstrap.Elasticsearch.execute(Elasticsearch.java:121) ~[elasticsearch-6.0.0.jar:6.0.0]
        at org.elasticsearch.cli.EnvironmentAwareCommand.execute(EnvironmentAwareCommand.java:69) ~[elasticsearch-6.0.0.jar:6.0.0]
        at org.elasticsearch.cli.Command.mainWithoutErrorHandling(Command.java:134) ~[elasticsearch-6.0.0.jar:6.0.0]
        at org.elasticsearch.cli.Command.main(Command.java:90) ~[elasticsearch-6.0.0.jar:6.0.0]
        at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:92) ~[elasticsearch-6.0.0.jar:6.0.0]
        at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:85) ~[elasticsearch-6.0.0.jar:6.0.0]
Caused by: java.lang.IllegalStateException: Unable to access 'path.data' (/usr/share/elasticsearch/data)
        at org.elasticsearch.bootstrap.FilePermissionUtils.addDirectoryPath(FilePermissionUtils.java:70) ~[elasticsearch-6.0.0.jar:6.0.0]
        at org.elasticsearch.bootstrap.Security.addFilePermissions(Security.java:287) ~[elasticsearch-6.0.0.jar:6.0.0]
        at org.elasticsearch.bootstrap.Security.createPermissions(Security.java:242) ~[elasticsearch-6.0.0.jar:6.0.0]
        at org.elasticsearch.bootstrap.Security.configure(Security.java:120) ~[elasticsearch-6.0.0.jar:6.0.0]
        at org.elasticsearch.bootstrap.Bootstrap.setup(Bootstrap.java:207) ~[elasticsearch-6.0.0.jar:6.0.0]
        at org.elasticsearch.bootstrap.Bootstrap.init(Bootstrap.java:322) ~[elasticsearch-6.0.0.jar:6.0.0]
        at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:130) ~[elasticsearch-6.0.0.jar:6.0.0]
        ... 6 more
Caused by: java.nio.file.AccessDeniedException: /usr/share/elasticsearch/data
        at sun.nio.fs.UnixException.translateToIOException(UnixException.java:84) ~[?:1.8.0_151]
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102) ~[?:1.8.0_151]
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107) ~[?:1.8.0_151]
        at sun.nio.fs.UnixFileSystemProvider.createDirectory(UnixFileSystemProvider.java:384) ~[?:1.8.0_151]
        at java.nio.file.Files.createDirectory(Files.java:674) ~[?:1.8.0_151]
        at java.nio.file.Files.createAndCheckIsDirectory(Files.java:781) ~[?:1.8.0_151]
        at java.nio.file.Files.createDirectories(Files.java:767) ~[?:1.8.0_151]
        at org.elasticsearch.bootstrap.Security.ensureDirectoryExists(Security.java:401) ~[elasticsearch-6.0.0.jar:6.0.0]
        at org.elasticsearch.bootstrap.FilePermissionUtils.addDirectoryPath(FilePermissionUtils.java:68) ~[elasticsearch-6.0.0.jar:6.0.0]
        at org.elasticsearch.bootstrap.Security.addFilePermissions(Security.java:287) ~[elasticsearch-6.0.0.jar:6.0.0]
        at org.elasticsearch.bootstrap.Security.createPermissions(Security.java:242) ~[elasticsearch-6.0.0.jar:6.0.0]
        at org.elasticsearch.bootstrap.Security.configure(Security.java:120) ~[elasticsearch-6.0.0.jar:6.0.0]
        at org.elasticsearch.bootstrap.Bootstrap.setup(Bootstrap.java:207) ~[elasticsearch-6.0.0.jar:6.0.0]
        at org.elasticsearch.bootstrap.Bootstrap.init(Bootstrap.java:322) ~[elasticsearch-6.0.0.jar:6.0.0]
        at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:130) ~[elasticsearch-6.0.0.jar:6.0.0]
        ... 6 more
```

Why is `path.data` being used? and why hasn't the installer handle permissions for this folder (`/usr/share/elasticsearch` is set as root:root)?

Easily reproducible using https://github.com/synhershko/elasticsearch-cloud-deploy/tree/es6 (I hit this with Azure but shouldn't matter). Note the installation scripts in ./packer and the userdata script in ./templates

**Comment by DaveCTurner** [comment]
Hi @synhershko! Your question is better suited to the support forums over at https://discuss.elastic.co. We prefer to use Github issues only for bug reports and feature requests, and we think it's more likely this is a question (about a nonstandard installer) than a bug report. I'm closing this as an issue here, but encourage you to ask this question in the forum instead. Many thanks.

**Comment by synhershko** [comment]
If you actually read the issue you'd see this is actually happening with the default installer provided and released by you.

**Comment by synhershko** [comment]
And also, a client-only node shouldn't be checking on the data path anyway. If that's not a bug I don't know what is.

**Comment by rjernst** [comment]
This is a "bug", of sorts. When configuring the security policy, we do not look at any settings to determine if the path may or may not be used. I'm going to reopen this, but know that this is very low priority.

**Comment by synhershko** [comment]
@rjernst if you can point me at the right file in the code base I'll be happy to adopt.

**Comment by DaveCTurner** [comment]
My apologies @synhershko, I misunderstood the origin of the installer.

**Comment by rjernst** [comment]
@synhershko `core/src/main/java/org/elasticsearch/bootstrap/Security.java`.

**Comment by jasontedor** [comment]
> If you actually read the issue you'd see this is actually happening with the default installer provided and released by you.

> If that's not a bug I don't know what is.

There is a bug here and the issue was closed prematurely which we will work on our end, but this kind of tone is still not acceptable. We are collaborators here, not adversaries.

**Comment by jasontedor** [comment]
For now you can work around this by simply creating that directory with permissions for the Elasticsearch user.

**Comment by synhershko** [comment]
> There is a bug here and the issue was closed prematurely which we will work on our end, but this kind of tone is still not acceptable. We are collaborators here, not adversaries.

You are right @jasontedor , and my apologies for that @DaveCTurner.

I'm going to look at that file and see if I can send a PR to fix it. I'd rather avoid creating that folder because I'm using automation scripts and those tend to get too specific too fast otherwise.

**Comment by jasontedor** [comment]
@synhershko I am not sure if I want that file to become node-type aware so I don’t think there’s an obvious fix at the moment. That might be the fix we have to go with but I want to see if we have other options first. 

Another workaround is to set `path.data` to `/var/lib/elasticsearch` or another directory like that.

**Comment by jasontedor** [comment]
> Another workaround is to set `path.data` to `/var/lib/elasticsearch` or another directory like that.

To be clear here: this is what the default packaging does and it ensures that directory exists.

**Comment by synhershko** [comment]
Okay, the fix is rather trivial other than the concern you raised:

```java
        // only need to access default/configured data paths if we are running on a data node
        if (Node.NODE_DATA_SETTING.get(environment.settings())) {
            if (environment.sharedDataFile() != null) {
                addDirectoryPath(policy, Environment.PATH_SHARED_DATA_SETTING.getKey(), environment.sharedDataFile(),
                    "read,readlink,write,delete");
            }
            final Set<Path> dataFilesPaths = new HashSet<>();
            for (Path path : environment.dataFiles()) {
                addDirectoryPath(policy, Environment.PATH_DATA_SETTING.getKey(), path, "read,readlink,write,delete");
            /*
             * We have to do this after adding the path because a side effect of that is that the directory is created; the Path#toRealPath
             * invocation will fail if the directory does not already exist. We use Path#toRealPath to follow symlinks and handle issues
             * like unicode normalization or case-insensitivity on some filesystems (e.g., the case-insensitive variant of HFS+ on macOS).
             */
                try {
                    final Path realPath = path.toRealPath();
                    if (!dataFilesPaths.add(realPath)) {
                        throw new IllegalStateException("path [" + realPath + "] is duplicated by [" + path + "]");
                    }
                } catch (final IOException e) {
                    throw new IllegalStateException("unable to access [" + path + "]", e);
                }
            }
            for (Path path : environment.repoFiles()) {
                addDirectoryPath(policy, Environment.PATH_REPO_SETTING.getKey(), path, "read,readlink,write,delete");
            }
        }
```

IMO this makes perfect sense - you only want to perform certain operations and throw exceptions on failure of them if the setup is right for it. I'm not sure why this wasn't happening in 5.x though - perhaps something in the installer changed? `/usr/share/elasticsearch` currently is owned by root:root.

Please let me know if I should submit a PR for it.

**Comment by synhershko** [comment]
> To be clear here: this is what the default packaging does and it ensures that directory exists.

@jasontedor How come 6.0 (using the deb installer) looks at `/usr/share/elasticsearch` ?

**Comment by jasontedor** [comment]
> How come 6.0 (using the deb installer) looks at `/usr/share/elasticsearch` ?

It looks at `path.data` which is packaged to be `/var/lib/elasticsearch`. You have overridden this to not be set so it falls back to the default which is `ES_HOME/data`. Here, `ES_HOME` is `/usr/share/elasticsearch`.

**Comment by synhershko** [comment]
Ok, so that is the change between 5.x and 6.x, thanks. Was that made on
purpose?

--

Itamar Syn-Hershko
Freelance Developer & Consultant
Elasticsearch Partner
Microsoft MVP | Lucene.NET PMC
http://code972.com | @synhershko <https://twitter.com/synhershko>
http://BigDataBoutique.co.il/

**Comment by jasontedor** [comment]
Indirectly, yes. I think the relevant change is the removal of default settings (e.g., `default.path.data`, see #25408). The point is that in the past if you did not set `path.data` then we would fall back to `default.path.data`. The service scripts set `default.path.data` to `/var/lib/elasticsearch` which again the packages ensure exists so there's no problem when `path.data` is not set. Now when `path.data` is not set we fall back to `ES_HOME/data` instead which is different than the previous fallback.

> Was that made on purpose?

Since it's not clear to me we are talking about the same change until I clarified which change it is, I can't answer this question. What are you asking?

**Comment by jasontedor** [comment]
@synhershko The fix you have proposed is not sufficient, it does not cover the case of master nodes. It also does not cover the case of coordinating-only nodes that do persist cluster state (`node.local_storage` defaults to true). And anyway, it makes this class node-aware which I do not want to do. I think I have a solution that I am happy with. I will open a PR later today.

**Comment by jasontedor** [comment]
By the way, there is actually no bug in *your* case because you have not set `node.local_storage` to false (defaults to true). There is a bug in the case that `node.local_storage` is false. In this case we should not be trying to access `path.data` but I would go further and say that we should reject the settings `node.local_storage` is false and `path.data` is set.

**Comment by jasontedor** [comment]
I opened #27587.

**Comment by synhershko** [comment]
> By the way, there is actually no bug in your case because you have not set node.local_storage to false (defaults to true).

Sorry this should have been a comment here - I do want `node.local_storage` to be set to true on my client-only nodes, to persist cluster state (because, why not).

**Comment by jasontedor** [comment]
> Sorry this should have been a comment here - I do want `node.local_storage` to be set to true on my client-only nodes, to persist cluster state (because, why not).

Right, and like I say: there's no bug here for you then because we are going to ensure that `path.data` exists and we have permissions for them in this case then. You have a broken configuration, you need to set `path.data` to `/var/lib/elasticsearch` (as the packaging defaults to and ensure exists) or somewhere else where you are comfortable ensuring Elasticsearch can store data.

The only bug is when `node.local_storage` is false and `node.master` and `node.data` are false. That's addressed by #27587 now.
